Sources/AppcuesKit/Presentation/Actions/ActionRegistry.swift

```diff
@@ -42,16 +47,31 @@ internal class ActionRegistry {
         actions[action.type] = action
     }
 
-    func actionClosures(for actionModels: [Experience.Action]) -> [(@escaping Completion) -> Void] {
-        guard let appcues = appcues else { return [] }
+    private func processFirstAction() {
+        guard !isProcessing, let appcues = appcues else { return }
 
-        return actionModels.compactMap { actionModel in
-            if let actionInstance = actions[actionModel.type]?.init(config: actionModel.config) {
-                return { completion in
-                    actionInstance.execute(inContext: appcues, completion: completion)
+        if let actionInstance = actionQueue.first {
+            isProcessing = true
+            actionInstance.execute(inContext: appcues) {
+                DispatchQueue.main.async {
+                    // On completion, remove the action, which triggers the didSet to process the remaining action handlers.
+                    self.isProcessing = false
+                    self.actionQueue.removeFirst()
```

Comment:
always looking for any race condition possibility in things like this 🤔

this _seems_ safe, since if something is enqueued while isProcessing == true, it would then get picked up and ran after the removeFirst.  

thinking out loud, but I don't see a way something will get left lingering here 👍 Kotlin made this easy with built in Channels.

edit: and you tested it well 👍

Reply:
Always a good question to raise! You've got it exactly right. Only removing from the start and adding to the end makes it simpler. And without the `isProcessing` check, adding a new action while an existing one was running would also trigger the didSet, meaning we'd be processing twice and the last one to finish would crash trying to `removeFirst` when empty.